tests/MongoDB.Driver.Tests/Specifications/Runner/BaseTestRunner.cs

```diff
@@ -0,0 +1,391 @@
+        {
+            var factory = new JsonDrivenTestFactory(client, DatabaseName, CollectionName, objectMap);
+
+            foreach (var operation in test[OperationsKey].AsBsonArray.Cast<BsonDocument>())
+            {
+                var receiver = operation["object"].AsString;
+                var name = operation["name"].AsString;
+                var jsonDrivenTest = factory.CreateTest(receiver, name);
+
+                jsonDrivenTest.Arrange(operation);
+                if (test["async"].AsBoolean)
+                {
+                    jsonDrivenTest.ActAsync(CancellationToken.None).GetAwaiter().GetResult();
+                }
+                else
+                {
+                    jsonDrivenTest.Act(CancellationToken.None);
+                }
+                jsonDrivenTest.Assert();
+            }
+        }
+
+        protected abstract void TestBody(BsonDocument shared, BsonDocument test, EventCapturer eventCapturer);
+
+        protected virtual void TestInitialize()
+        {
+            // do nothing by default.
+        }
+
+        protected virtual void ValidateServerRequirements(BsonDocument document)
+        {
+            if (document.Contains("minServerVersion"))
+            {
+                var minServerVersion = document["minServerVersion"].AsString;
+                RequireServer.Check().VersionGreaterThanOrEqualTo(minServerVersion);
+            }
+        }
+
+        protected virtual void VerifyCollectionOutcome(BsonDocument outcome)
+        {
+            foreach (var aspect in outcome)
+            {
+                switch (aspect.Name)
+                {
+                    case "data":
+                        VerifyCollectionData(aspect.Value.AsBsonArray.Cast<BsonDocument>());
+                        break;
+
+                    default:
+                        throw new FormatException($"Unexpected collection outcome aspect: {aspect.Name}.");
+                }
+            }
+        }
+        #endregion
+
+        protected virtual void AssertEvents(EventCapturer actualEvents, BsonDocument test)
+        {
+            AssertEvents(actualEvents, test, null);
+        }
+
+        protected virtual void AssertEvents(EventCapturer actualEvents, BsonDocument test, Func<BsonDocument, BsonDocument> prepareEventResult)
+        {
+            if (test.Contains(ExpectationsKey))
+            {
+                var expectedEvents = test[ExpectationsKey].AsBsonArray.Cast<BsonDocument>().GetEnumerator();
+
+                while (actualEvents.Any())
+                {
+                    var actualEvent = actualEvents.Next();
+
+                    if (!expectedEvents.MoveNext())
+                    {
+                        throw new Exception($"Unexpected event of type: {actualEvent.GetType().Name}.");
+                    }
+
+                    var expectedEvent = expectedEvents.Current;
+                    prepareEventResult?.Invoke(expectedEvent);
```

Comment:
Done. 

>Note: we've gotten in trouble modifying the data that comes from the JSON files because many tests execute twice, once for async and once for sync. The second time the data has already been modified.

here, it will not affect us since we replace value only in event result for each test independently.